In [5]:
import numpy as np
import torch
import torch.nn as nn
import matplotlib.pyplot as plt

# 设置matplotlib为黑色主题（与课件保持一致）
plt.style.use('dark_background')
plt.rcParams.update({
    'figure.facecolor': 'black',
    'axes.facecolor': 'black',
    'axes.edgecolor': 'white',
    'axes.labelcolor': 'white',
    'text.color': 'white',
    'xtick.color': 'white',
    'ytick.color': 'white',
    'grid.color': '#444444',
    'grid.alpha': 0.3,
    'legend.facecolor': 'black',
    'legend.edgecolor': 'white',
})

print('✅ 环境准备完成！')

✅ 环境准备完成！


# <font color="#00FFFF" > 多层感知机 - 课堂练习</font>

本练习包含三个部分，分别对应课件中的核心知识点：
- **练习1**：概率与Logit的相互转换（Sigmoid函数理解）
- **练习2**：Softmax手动计算（多分类概率理解）

---

## <font color="#FFEA00" >练习1：概率与Logit的相互转换</font>

### 📚 背景知识回顾

**Logit函数**（对数几率）：将概率 p ∈ (0, 1) 映射到实数 z

$$z = logit(p) = log(\frac{p}{1-p})$$

**Sigmoid函数**：Logit的反函数，将实数 z 映射回概率 p

$$p = \sigma(z) = \frac{1}{1 + e^{-z}}$$

### 🎯 任务说明

请完成下表中的计算。可以使用计算器或Python辅助计算。

**提示**：
- e^x 可以用 `np.exp(x)` 计算
- log(x) 可以用 `np.log(x)` 计算（自然对数）
- 当 z = 0 时，sigmoid(0) = 0.5

### ✏️ 答题区域

| 序号 | 概率 p | Logit z = log(p/(1-p)) | Sigmoid验证 σ(z) |
|:---:|:---:|:---:|:---:|
| 1 | 0.5 | 0 | 0.5 |
| 2 | 0.8 | 2.079 | 0.8 |
| 3 | 0.2 | -1.386 | 0.2 |
| 4 | 0.5 | 0 | 1 |
| 5 | 0.881 | 2 | 0.881 |
| 6 | 0.269 | -1 | 0.269 |

---

In [6]:
# ==========================================
# 练习1：辅助计算区域（学生可使用）
# ==========================================

# 定义辅助函数
def logit(p):
    """从概率计算logit"""
    return np.log(p / (1 - p))

def sigmoid(z):
    """从logit计算概率"""
    return 1 / (1 + np.exp(-z))

# 示例：计算 p=0.75 对应的 logit
p_example = 0.75
z_example = logit(p_example)
p_back = sigmoid(z_example)

print(f'示例：p={p_example} -> logit={z_example:.4f} -> sigmoid={p_back:.4f}')

# TODO: 请在这里完成你的计算
# p_values = [0.5, 0.8, 0.2, ...]
# for p in p_values:
#     z = logit(p)
#     print(f'p={p} -> z={z:.4f}')

# TODO: 从logit反推概率
# z_values = [0, 2, -1]
# for z in z_values:
#     p = sigmoid(z)
#     print(f'z={z} -> p={p:.4f}')

示例：p=0.75 -> logit=1.0986 -> sigmoid=0.7500


---

## <font color="#FFEA00" >练习2：Softmax手动计算</font>

### 📚 背景知识回顾

**Softmax函数**：将 K 个logits转换为概率分布

$$softmax(z_i) = \frac{e^{z_i}}{\sum_{j=1}^{K} e^{z_j}}$$

**关键性质**：
- 所有输出概率之和为1
- 输入值越大，对应的概率越高
- 指数函数会放大差距（强者愈强）

### 🎯 任务说明

给定一个三分类问题的logits，手动计算Softmax概率。

**提示**：
- 先计算每个logit的指数 $e^{z_i}$
- 计算分母（所有指数之和）
- 每个类别的概率 = 该类的指数 / 分母

### ✏️ 任务1：基础Softmax计算

给定 logits: z = [2.0, 1.0, 0.1]

**步骤1**：计算每个logit的指数值（保留2位小数）

| 类别 | Logit z_i | $e^{z_i}$ |
|:---:|:---:|:---:|
| 0 | 2.0 | $e^{2.0} = 7.39$ |
| 1 | 1.0 | $e^{1.0} = 2.72$ |
| 2 | 0.1 | $e^{0.1} = 1.10$ |
| | **求和** | $7.39 + 2.72 + 1.10 = 11.21$ |

**步骤2**：计算Softmax概率

| 类别 | 计算过程 | Softmax概率 |
|:---:|:---|:---:|
| 0 | $e^{2.0}$ / sum | $7.39 / 11.21 = 0.66$ |
| 1 | $e^{1.0}$ / sum | $2.72 / 11.21 = 0.25$ |
| 2 | $e^{0.1}$ / sum | $1.10 / 11.21 = 0.10$ |
| | **概率之和** | $0.66 + 0.25 + 0.10 = 1.00$ |

**步骤3**：哪个类别被预测为正确类别？为什么？

**你的回答**：

```
第一个类别被预测为正确类别，因为它的Softmax概率最大（0.66）。



```

### ✏️ 任务2：观察温度效应

现在将原始logits都乘以10，即 z' = [20, 10, 1]

**步骤1**：完成下表

| 类别 | 原始Logit | 放大10倍 z' | $e^{z'}$ | Softmax概率 |
|:---:|:---:|:---:|:---:|:---:|
| 0 | 2.0 | 20 | $e^{20}$ | **≈ 1.000** |
| 1 | 1.0 | 10 | $e^{10}$ | **≈ 0.000** |
| 2 | 0.1 | 1 | $e^{1}$ | **≈ 0.000** |

**步骤2**：与任务1的结果对比，概率分布发生了什么变化？

**你的回答**：
有着比较大参数的类别（0）的Softmax概率最大，而其他类别（1和2）的Softmax概率都非常小。因为扩大倍数对他的增益最大
```
```

---

现在将原始logits都除以10，即 z'' = [0.2, 0.1, 0.01]

**步骤3**：完成下表

| 类别 | 原始Logit | 缩小10倍 z'' | $e^{z''}$ | Softmax概率 |
|:---:|:---:|:---:|:---:|:---:|
| 0 | 2.0 | 0.2 | $e^{0.2}$ | **0.375** |
| 1 | 1.0 | 0.1 | $e^{0.1}$ | **0.340** |
| 2 | 0.1 | 0.01 | $e^{0.01}$ | **0.285** |

**步骤4**：与任务1的结果对比，概率分布发生了什么变化？

**你的回答**：

```

不同类别之间的差异变小了，但是原来最大的softmax还是最大的。


```

In [7]:
# ==========================================
# 练习2：辅助计算区域（学生可使用）
# ==========================================

def softmax(z):
    """
    Softmax函数
    z: 一维数组，shape (n_classes,)
    """
    exp_z = np.exp(z - np.max(z))  # 数值稳定性技巧
    return exp_z / np.sum(exp_z)

# 任务1：基础计算
z1 = np.array([2.0, 1.0, 0.1])
print('=== 任务1：原始Logits ===')
print(f'Logits: {z1}')
print(f'Exp values: {np.exp(z1)}')
print(f'Sum of exp: {np.sum(np.exp(z1))}')
print(f'Softmax: {softmax(z1)}')
print(f'Sum of probs: {np.sum(softmax(z1))}')

# 任务2：放大10倍
z2 = z1 * 10
print('\n=== 任务2：放大10倍 ===')
print(f'Logits: {z2}')
# TODO: 完成计算

# 任务3：缩小10倍
z3 = z1 / 10
print('\n=== 任务3：缩小10倍 ===')
print(f'Logits: {z3}')
# TODO: 完成计算

=== 任务1：原始Logits ===
Logits: [2.  1.  0.1]
Exp values: [7.3890561  2.71828183 1.10517092]
Sum of exp: 11.212508845465344
Softmax: [0.65900114 0.24243297 0.09856589]
Sum of probs: 1.0

=== 任务2：放大10倍 ===
Logits: [20. 10.  1.]

=== 任务3：缩小10倍 ===
Logits: [0.2  0.1  0.01]


### 🤔 思考题

**问题1**：Softmax为什么叫Softmax？与Hardmax有什么区别？

**你的回答**：

```




```

---

**问题2**：在神经网络训练中，为什么温度（temperature）参数很重要？结合任务2和任务3的结果说明。

提示：温度参数 τ 的Softmax公式：$\operatorname{softmax}\left(z_{i} / \tau\right)=\frac{e^{z_{i} / \tau}}{\sum_{j} e^{z_{j} / \tau}}$

**你的回答**：

```
- 温度系数调整softmax的软硬程度 温度趋向于0时，softmax趋向于hardmax，温度趋向于无穷大时，softmax趋向于均匀分布，标准的时候温度为1。
- 温度参数的定义来自于玻尔兹曼分布，用于调整softmax的输出概率分布的平滑程度。
- 实验初期可以使用较高的温度，以便模型能够探索更多的分类可能性。
- 随着模型训练的进行，温度可以逐渐降低，使模型更专注于正确的分类。
- 这个就叫做模拟退火的训练方式
- 温度参数的调整可以在训练过程中动态调整，以平衡模型的探索和利用能力。
- 还有一个可以搭配温度系数的东西叫做top-p（核采样），选出一组概率和刚好达到p的最小次组合，选择动态数量的词，直到累计概率达到p，
Top-k：是“数个数”。不管概率多少，只看前 k 个词。缺点：如果前几个词概率都很低，AI 也会硬选，导致胡言乱语。Top-p：是“看质量”。根据概率分布动态调整选词范围。优点：更智能。当 AI 很有把握时，可选范围会缩得很小；当 AI 犹豫时，范围会自动扩大。
先调温度，发现重复僵化就升温，发现逻辑乱，幻觉多就降温，在动top-p，微调词汇池 p比较低的时候，相当于高精度筛选，输出集中安全，适合逻辑严谨的任务
```
